src/terminal/adapter/PageManager.cpp

```diff
@@ -0,0 +1,231 @@
+    const auto visibleTop = visibleViewport.top;
+    const auto wasVisible = _activePageNumber == _visiblePageNumber;
+    const auto newPageNumber = std::min(std::max(pageNumber, 1), MAX_PAGES);
+    auto redrawRequired = false;
+
+    // If we're changing the visible page, what we do is swap out the current
+    // visible page into its backing buffer, and swap in the new page from the
+    // backing buffer to the main buffer. That way the rest of the system only
+    // ever has to deal with the main buffer.
+    if (makeVisible && _visiblePageNumber != newPageNumber)
+    {
+        const auto& newBuffer = _getBuffer(newPageNumber, pageSize);
+        auto& saveBuffer = _getBuffer(_visiblePageNumber, pageSize);
+        for (auto i = 0; i < pageSize.height; i++)
+        {
+            saveBuffer.GetMutableRowByOffset(i).CopyFrom(visibleBuffer.GetRowByOffset(visibleTop + i));
+        }
+        for (auto i = 0; i < pageSize.height; i++)
+        {
+            visibleBuffer.GetMutableRowByOffset(visibleTop + i).CopyFrom(newBuffer.GetRowByOffset(i));
+        }
+        _visiblePageNumber = newPageNumber;
+        redrawRequired = true;
+    }
+
+    // If the active page was previously visible, and is now still visible,
+    // there is no need to update any buffer properties, because we'll have
+    // been using the main buffer in both cases.
+    const auto isVisible = newPageNumber == _visiblePageNumber;
+    if (!wasVisible || !isVisible)
+    {
+        // Otherwise we need to copy the properties from the old buffer to the
+        // new, so we retain the current attributes and cursor position. This
+        // is only needed if they are actually different.
+        auto& oldBuffer = wasVisible ? visibleBuffer : _getBuffer(_activePageNumber, pageSize);
+        auto& newBuffer = isVisible ? visibleBuffer : _getBuffer(newPageNumber, pageSize);
+        if (&oldBuffer != &newBuffer)
+        {
+            // When copying the cursor position, we need to adjust the y
+            // coordinate to account for scrollback in the visible buffer.
+            const auto oldTop = wasVisible ? visibleTop : 0;
+            const auto newTop = isVisible ? visibleTop : 0;
+            auto position = oldBuffer.GetCursor().GetPosition();
+            position.y = position.y - oldTop + newTop;
+            newBuffer.SetCurrentAttributes(oldBuffer.GetCurrentAttributes());
+            newBuffer.CopyProperties(oldBuffer);
+            newBuffer.GetCursor().SetPosition(position);
+        }
+        // If we moved from the visible buffer to a background buffer
+        // we need to hide the cursor in the visible buffer. It should
+        // be restored when moving back.
+        if (wasVisible && !isVisible)
+        {
+            visibleBuffer.GetCursor().SetIsVisible(false);
+        }
+    }
+
+    _activePageNumber = newPageNumber;
+    if (redrawRequired)
+    {
+        _renderer.TriggerRedrawAll();
+    }

```

Comment:
Would it make sense to give the caller the responsibility to call the renderer, so that we can decouple these components better?

Reply:
I did actually try that at one point, but it wasn't very clean, because the redraw then needs to be handled in a bunch of different places. And it doesn't decouple the renderer from the `PageManager` anyway, because we still need the renderer for the buffer creation.